### PR TITLE
Bfme: fixes, RadialMenu and BuildingPlots

### PR DESCRIPTION
**Pull request by Tarcontar**
- Bfme castles are unpacked
- Buildings can be build on BuildPlots
- Made the RadialMenu also function in Bfme
- fixed modelConditionStates and idle animations

### CODE REVIEW

File: src/OpenSage.Game/Logic/Object/Behaviors/CastleBehavior.cs [hunk 1]
```diff
@@ -65,9 +65,10 @@ private void Unpack(GameContext context)
         internal override void Update(BehaviorUpdateContext context)
         {
             // TODO: Figure out the other unpack conditions
-            if (_moduleData.InstantUnpack)
+            //if (_moduleData.InstantUnpack)
```
Comment (Tarcontar) [inline src/OpenSage.Game/Logic/Object/Behaviors/CastleBehavior.cs:68]: TODO: how to unpack the Bfme castles

File: src/OpenSage.Game/Logic/Object/Behaviors/CastleBehavior.cs [hunk 2]
```diff
@@ -83,9 +84,11 @@ private void AssignOwner(GameObject gameObject)
         private CastleEntry FindCastle()
         {
             // Use the gameobject side
+            return _moduleData.CastleToUnpackForFactions[0];
```
Comment (Tarcontar) [inline src/OpenSage.Game/Logic/Object/Behaviors/CastleBehavior.cs:87]: TODO: handle this properly

File: src/OpenSage.Game/Scene3D.cs [hunk 3]
```diff
@@ -108,7 +108,15 @@ internal Scene3D(Game game, MapFile mapFile, string mapPath, int randomSeed)
...
-            LocalPlayer = _players.FirstOrDefault();
+            var valids = _players.Where(x => x.Side.Contains("Gondor") || x.Side.Contains("USA"));
+            if (valids.Count() > 0)
+            {
+                LocalPlayer = valids.First();
+            }
+            else
+            {
+                LocalPlayer = _players.FirstOrDefault();
+            }
```
Comment (Tarcontar) [inline src/OpenSage.Game/Scene3D.cs:119]: hack to get a valid faction when jumping to a map in Bfme -> .First() is Civilian and thus no castle will be unpacked

File: src/OpenSage.Game/Logic/OrderGenerators/ConstructBuildingOrderGenerator.cs [hunk 4]
```diff
@@ -76,11 +76,11 @@ public void BuildRenderList(RenderList renderList, Camera camera, in TimeInterva
         public OrderGeneratorResult TryActivate(Scene3D scene, KeyModifiers keyModifiers)
         {
             // TODO: Probably not right way to get dozer object.
-            var dozer = scene.LocalPlayer.SelectedUnits.First();
+            //var dozer = scene.LocalPlayer.SelectedUnits.First();
```
Comment (Tarcontar) [inline src/OpenSage.Game/Logic/OrderGenerators/ConstructBuildingOrderGenerator.cs:79]: TODO: find a pendant for bfme and later games

File: src/OpenSage.Game/Logic/Object/Behaviors/CastleBehavior.cs (hunk 1 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Logic/Object/Behaviors/CastleBehavior.cs]: fixed

File: src/OpenSage.Game/Logic/Object/Behaviors/CastleBehavior.cs (hunk 2 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Logic/Object/Behaviors/CastleBehavior.cs]: fixed

File: src/OpenSage.Game/Scene3D.cs (hunk 3 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Scene3D.cs]: fixed

File: src/OpenSage.Game/Logic/OrderGenerators/ConstructBuildingOrderGenerator.cs (hunk 4 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Logic/OrderGenerators/ConstructBuildingOrderGenerator.cs]: fixed

File: src/OpenSage.Game/Audio/AudioSystem.cs [hunk 5]
```diff
@@ -185,9 +185,13 @@ private void UpdateListener(Camera camera)
 
         public void PlayAudioEvent(GameObject emitter, BaseAudioEventInfo baseAudioEvent)
         {
+            return;
```
Comment (Tarcontar) [inline src/OpenSage.Game/Audio/AudioSystem.cs:188]: Bfme crashes at the moment when some specific units are selected, @feliwir will invesitgate this further.

Comment (valentinhimself) [reply src/OpenSage.Game/Audio/AudioSystem.cs:188]: > Bfme crashes at the moment when some specific units are selected, @feliwir will invesitgate this further.

Crashes are likely to be related to their commandsets and specifically to certain commandbuttons that those specific units have.

Comment (feliwir) [reply src/OpenSage.Game/Audio/AudioSystem.cs:188]: No, it’s due to the 3d soundsystem, so @Tarcontar is correct

Comment (feliwir) [reply src/OpenSage.Game/Audio/AudioSystem.cs:188]: @Tarcontar Could you make this call the regular **PlayAudioEvent** for now and add a TODO please? I'll investigate once i have time

File: src/OpenSage.Game/Gui/ControlBar/CommandButton.cs [hunk 6]
```diff
@@ -8,6 +8,15 @@ namespace OpenSage.Gui.ControlBar
 {
     public sealed class CommandButton : BaseAsset
     {
+        public CommandButton(){ }
```
Comment (feliwir) [inline src/OpenSage.Game/Gui/ControlBar/CommandButton.cs:11]: I don't think this is valid OpenSage formatting

Comment (Tarcontar) [reply src/OpenSage.Game/Gui/ControlBar/CommandButton.cs:11]: done

File: src/OpenSage.Game/Logic/Object/GameObject.cs [hunk 7]
```diff
@@ -403,6 +405,19 @@ internal void LogicTick(ulong frame, in TimeInterval time)
             }
         }
 
+        public bool CanRecruitHero(ObjectDefinition definition)
+        {
+            foreach (var obj in Parent.Items)
+            {
+                if (obj.Definition.Name == definition.Name
```
Comment (tgjones) [inline src/OpenSage.Game/Logic/Object/GameObject.cs:412]: Why can we not compare object references? I thought we had a unique `ObjectDefinition` instance per name.

File: src/OpenSage.Game/Logic/Object/Update/ProductionExit/QueueProductionExitUpdate.cs [hunk 8]
```diff
@@ -16,7 +16,7 @@ internal QueueProductionExitUpdate(QueueProductionExitUpdateModuleData moduleDat
 
         Vector3? IProductionExit.GetNaturalRallyPoint() => _moduleData.NaturalRallyPoint;
 
-        public int ExitDelay => _moduleData.ExitDelay; 
+        public int ExitDelay => _moduleData.ExitDelay > 0 ? _moduleData.ExitDelay : 300;
```
Comment (tgjones) [inline src/OpenSage.Game/Logic/Object/Update/ProductionExit/QueueProductionExitUpdate.cs:19]: If this is the default value, I think it should be on the property declaration [here](https://github.com/OpenSAGE/OpenSAGE/blob/aabedaf9537749f9e522f8cf97f5ad3085aca52f/src/OpenSage.Game/Logic/Object/Update/ProductionExit/QueueProductionExitUpdate.cs#L48).

File: src/OpenSage.Game/Logic/Object/Update/ProductionUpdate.cs [hunk 9]
```diff
@@ -232,6 +232,18 @@ private void GrantUpgrade(UpgradeTemplate upgradeDefinition)
 
         internal bool CanProduceObject(ObjectDefinition objectDefinition)
         {
+            // only one hero of the same kind can be produced at a time
+            if (objectDefinition.KindOf.Get(ObjectKinds.Hero))
+            {
+                foreach (var job in ProductionQueue)
+                {
+                    if (job.ObjectDefinition.Name == objectDefinition.Name)
```
Comment (tgjones) [inline src/OpenSage.Game/Logic/Object/Update/ProductionUpdate.cs:240]: Same here - do we not have referential equality for `ObjectDefinition` instances?

**Review by tgjones — APPROVED** [review]
LGTM 👍

File: src/OpenSage.Game/Logic/Object/Update/ProductionExit/QueueProductionExitUpdate.cs (hunk 8 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Logic/Object/Update/ProductionExit/QueueProductionExitUpdate.cs:19]: fixed

File: src/OpenSage.Game/Logic/Object/Update/ProductionUpdate.cs (hunk 9 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Logic/Object/Update/ProductionUpdate.cs:240]: fixed

File: src/OpenSage.Game/Logic/Object/GameObject.cs (hunk 7 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Logic/Object/GameObject.cs:412]: done

File: src/OpenSage.Game/Logic/Object/Update/ProductionExit/QueueProductionExitUpdate.cs (hunk 8 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Logic/Object/Update/ProductionExit/QueueProductionExitUpdate.cs:19]: just checked, the corresponding building explicitly sets it to 0, so i removed the default value of 300 alltogether

**Pull request merged**